Spigot-Server-Patches/0381-Optimize-redstone-algorithm.patch

```diff
@@ -0,0 +1,1185 @@
+From 6a5104193614fbe321dca74d32646a4ad33595fa Mon Sep 17 00:00:00 2001
+From: egg82 <phantom_zero@ymail.com>
+Date: Thu, 27 Sep 2018 01:43:35 -0600
+Subject: [PATCH] Optimize redstone algorithm
+
+This patch implements theosib's redstone algorithms to completely overhaul the way redstone works.
+The new algorithms should be many times faster than current vanilla ones.
+From the original author's comments, it looks like it shouldn't interfere with any redstone save for very extreme edge-cases.
+
+Surprisingly, not a lot was touched aside from a few obfuscation helpers and BlockRedstoneWire.
+A lot of this code is self-contained in a helper class.
+
+Aside from making the obvious class/function renames and obfhelpers I didn't need to modify much.
+Just added Bukkit's event system and took a few liberties with dead code and comment misspellings.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index ce5273384..1035dfd43 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -388,4 +388,14 @@ public class PaperConfig {
+             log("Async Chunks: Enabled - Chunks will be loaded and generated much faster, without lag.");
+         }
+     }
++
++    public static boolean useNewRedstone = false;
++    private static void useNewRedstone() {
++        useNewRedstone = config.getBoolean("settings.use-new-redstone", false);
```

Comment:
IMO a better name for this setting would be `use-theosib-redstone`.

There are a number of alternate redstone algorithms, in particular the Panda4994 system - which is quite popular, and available as a Bukkit plugin, Forge mod & Sponge config option. Just using "new redstone" doesn't really describe the feature very clearly.

In fact, perhaps a string value would make more sense, in case another algorithm is added in the future.

e.g.

```yaml
redstone-algorithm: vanilla
redstone-algorithm: theosib
redstone-algorithm: panda
```

Reply:
Yup, latter is better and easier to implement as well

Reply:
Better name definitely needed, but as discussed in discord we weren’t planning on implementing pandas algorithm

Reply:
Yeah, the name can probably be better. I'm terrible at those. I suggested implementing something similar to what @luck was saying because I had spent most of the day trying to port Sponge's version of Panda's algorithm but it's too buggy, touches too many files, and overall this one is better and likely faster.

Reply:
Even if there's no current intention of porting Panda's, another system may pop up in the future that's worthwhile.

No harm in using a generic setting (my second example above) just to ease that process down the line.

Reply:
Hi, everyone.  

I'm very flattered that my redstone accelerator was chosen for inclusion in Paper.  I was asked about naming, and my suggestion would be to call the setting "use-eigencraft-redstone".  It is a higher priority for me to draw attention to the EigenCraft *community* than to make sure everyone sees my pseudonym or whatever.

Thanks.